app/components/RichEditor/slatePlugin.js

```diff
@@ -263,103 +201,73 @@ const plugin = {
+            ? match + INDENT
+            : match
         )
-        const firstPart = parentBlocks
-          .slice(0, higherLevelItemIndex)
-          .filter(({key}) => document.getDepth(key) === firstItemDepth)
-        const secondPart = parentBlocks
-          .slice(higherLevelItemIndex)
-          .filter(({key}) => document.getDepth(key) === higherLevelItemDepth)
-
-        // Only proceed if all items can be moved.
-        if (
-          canIndentItems(editor, firstPart) &&
-          canIndentItems(editor, secondPart)
-        ) {
-          indentListItems(editor, firstPart)
-          indentListItems(editor, secondPart)
-        }
+        const lengthDiff = newText.length - text.length
+        const newAnchorOffset = selection.isForward
+          ? startOffset + INDENT.length
+          : endOffset + lengthDiff
+        const newFocusOffset = selection.isBackward
+          ? startOffset + INDENT.length
+          : endOffset + lengthDiff
+
+        editor
+          // `setNodeByKey` doesn't work with changing text. Implementation using
+          // `insertText` (preserves selection offsets even through undos) turned
+          // out to be too complex.
+          .replaceNodeByKey(key, {object: 'text', text: newText})
+          .moveAnchorTo(path, newAnchorOffset)
+          .moveFocusTo(path, newFocusOffset)
       } else {
-        indentListItems(
-          editor,
-          parentBlocks.filter(
-            ({key}) => document.getDepth(key) === firstItemDepth
-          )
-        )
+        changeListIndentation(editor, 'increase')
       }
-
-      editor
-        .moveAnchorTo(anchorKey, anchorOffset)
-        .moveFocusTo(focusKey, focusOffset)
     },
     deindent(editor) {
       const {blocks, document, selection} = editor.value
-      const {anchor, focus} = selection
-      const {key: anchorKey, offset: anchorOffset} = anchor
-      const {key: focusKey, offset: focusOffset} = focus
 
-      const parentBlocks = blocks
-        .map(({key}) => document.getParent(key))
-        .skipUntil(isListItemNode)
-        .takeWhile(isListItemNode)
-
-      if (!parentBlocks.size) return
-
-      const firstItemDepth = document.getDepth(parentBlocks.first().key)
-      const higherLevelItemIndex = parentBlocks.findIndex(
-        ({key}) => document.getDepth(key) < firstItemDepth
-      )
-
-      if (higherLevelItemIndex !== -1) {
-        const higherLevelItemDepth = document.getDepth(
-          parentBlocks.get(higherLevelItemIndex).key
+      if (editor.isInBlocks(Nodes.BLOCK_CODE) && blocks.size === 1) {
+        const INDENT = '  '
+        const {start, end} = selection
+        const {key, offset: startOffset} = start
+        const {offset: endOffset} = end
+        const {text} = document.getNode(key)
+        const path = document.getPath(key)
+        const startIndex = text.lastIndexOf('\n', startOffset - 1)
+        const re = RegExp(`(^|\n)${INDENT}`, 'g')
```

Comment:
Nitpick: could you rename this to `regularExpression` or `expression` or something? Not a fan of acronyms as variable names.